File: apple/Sources/SignetAppleCore/AccessControl.swift
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2026 Nirapod Labs

/// The access-control policy requested for a key at generation. On Apple it is
/// applied as the key's `SecAccessControl`; the achieved `AuthClass` is reported
/// in `SecurityTierReport.authEnforced`.
public struct AccessControlPolicy: Sendable, Equatable {
    /// The presence check requested for the key.
    public enum AuthRequirement: String, Sendable, Equatable, CaseIterable {
        case none
        case biometricOnly
        case biometricOrDeviceCredential
    }

    /// The requested presence check.
    public let authRequirement: AuthRequirement
    /// Auth reuse window. `nil` or `0` means per-use auth; `> 0` means a time
    /// window. On Apple it is a sign-time value, not baked into the key; the
    /// current sign path does not apply it.
    public let authValiditySeconds: Int?
    /// Whether a biometric re-enrollment invalidates the key. `true` maps to
    /// `.biometryCurrentSet`, `false` to `.biometryAny`.
    public let invalidateOnBiometricEnrollment: Bool

    public init(
        authRequirement: AuthRequirement = .none,
        authValiditySeconds: Int? = nil,
        invalidateOnBiometricEnrollment: Bool = true
    ) {
        self.authRequirement = authRequirement
        self.authValiditySeconds = authValiditySeconds
        self.invalidateOnBiometricEnrollment = invalidateOnBiometricEnrollment
    }

    /// No presence check: usable without biometric or passcode auth.
    public static let none = AccessControlPolicy(authRequirement: .none)
}

File: kmp/signet/src/commonMain/kotlin/xyz/nirapod/signet/kmp/AccessControl.kt
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2026 Nirapod Labs

package xyz.nirapod.signet.kmp

/** The presence check a caller requests on a key. */
public enum class AuthRequirement {
    none,
    biometricOnly,
    biometricOrDeviceCredential,
}

/**
 * Access-control policy for a key. `authValiditySeconds` null or 0 means per-use
 * auth; a positive value is a time window. `invalidateOnBiometricEnrollment`
 * (default true) invalidates the key when the biometric set changes.
 */
public class AccessControlPolicy(
    public val authRequirement: AuthRequirement,
    public val authValiditySeconds: Int? = null,
    public val invalidateOnBiometricEnrollment: Boolean = true,
) {
    public companion object {
        /** No presence check. */
        public val None: AccessControlPolicy = AccessControlPolicy(AuthRequirement.none)
    }
}
